ES|QL: Add TBUCKET function

leontyevdv opened the pull request:
Introduce the function TBUCKET(<time interval>) which applies grouping on the @timestamp field, truncating its value to the specified granularity:

TBUCKET(1h) is equivalent to BUCKET(1 hour, @timestamp) TBUCKET(7d) is equivalent to BUCKET(7 days, @timestamp)

Closes #131068

leontyevdv committed:
ES|QL: Add TBUCKET function

Introduce the function TBUCKET(<time interval>) which applies grouping
on the @timestamp field, truncating its value to the specified
granularity:

TBUCKET(1h) is equivalent to BUCKET(1 hour, @timestamp)
TBUCKET(7d) is equivalent to BUCKET(7 days, @timestamp)

Closes #131068

leontyevdv committed:
ES|QL: Add TBUCKET function

Introduce the function TBUCKET(<time interval>) which applies grouping
on the @timestamp field, truncating its value to the specified
granularity:

TBUCKET(1h) is equivalent to BUCKET(1 hour, @timestamp)
TBUCKET(7d) is equivalent to BUCKET(7 days, @timestamp)

Closes #131068

leontyevdv committed:
Merge branch 'main' into feature/esql-tbucket-function

fang-xing-esql commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/grouping/TBucket.java:
```diff
@@ -0,0 +1,168 @@
+    implements
+        PostOptimizationVerificationAware,
+        TwoOptionalArguments,
+        LocalSurrogateExpression {
```
There is another interface [SurrogateExpression](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/SurrogateExpression.java), that does the substitution at coordinator node, I wonder if there is a particular reason that this rewrite needs to be done at data node?

leontyevdv replied:
Thanks for pointing that out. I was unaware of the different types of execution, e.g., on the coordinator/data nodes. This is my first task at Elastic, and I'm still figuring out how everything works. I implemented the SurrogateExpression interface to make the substitution work on the coordinator node.

leontyevdv committed:
ES|QL: Add TBUCKET function

Replace evaluation by a surrogate.

Closes #131068

leontyevdv committed:
ES|QL: Add TBUCKET function

Replace evaluation by a surrogate.

Closes #131068

leontyevdv committed:
ES|QL: Add TBUCKET function

Fix tests

Closes #131068

leontyevdv commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/grouping/TBucket.java:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    @Override
+    public ExpressionEvaluator.Factory toEvaluator(ToEvaluator toEvaluator) {
```
The initial idea was to create a new function that would work directly with DateTrunc and bypass Bucket altogether. Perhaps, it was the wrong idea. With the surrogate it works perfectly.

leontyevdv committed:
Merge branch 'main' into feature/esql-tbucket-function

leontyevdv committed:
ES|QL: Add TBUCKET function

Cleanup tests

Closes #131068

leontyevdv committed:
ES|QL: Add TBUCKET function

Add TBUCKET capability

Closes #131068

leontyevdv committed:
Merge branch 'main' into feature/esql-tbucket-function

leontyevdv committed:
ES|QL: Add test for TBucket into the AnalyzerTests

Closes #131068

not-napoleon reviewed (commented):
We need some CSV tests for error cases.  What happens if you use `TBUCKET` on a data set with no `@timezone` field, for example?  Presumably that fails in some graceful way.  What other edge cases can we test for?

I think the implementation and the abstract function tests are fine (although I'm shocked we don't have randomized periods and durations in the test framework, but fixing that is beyond the scope of this PR)

leontyevdv committed:
ES|QL: Address PR review comments

Remove skip version from tbucket.csv-spec
Polish code in AnalyzerTests

Closes #131068

leontyevdv commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/tbucket.csv-spec:
```diff
@@ -0,0 +1,204 @@
+required_capability: tbucket
+
+FROM sample_data
+| STATS min = MAX(@timestamp), max = MAX(@timestamp) BY bucket = TBUCKET(10 seconds)
```
This was a mistake and it's been fixed. I took an example from https://github.com/elastic/elasticsearch/blob/748575096a0544748f98d119f048499c4dbee791/x-pack/plugin/esql/qa/testFixtures/src/main/resources/bucket.csv-spec#L289. Should we maybe fix it too?

not-napoleon replied:
🤦🏼‍♀️  ...yes, we should fix that too.  If you wouldn't mind opening a second PR to do that when you have a chance, I'd appreciate it.

leontyevdv commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/AnalyzerTests.java:
```diff
@@ -4132,6 +4133,41 @@ public void testGroupingOverridesInInlinestats() {
+            | SORT min
+            """, "mapping-sample_data.json");
+
+        var limit = as(plan, Limit.class);
```
This style is used pretty much everywhere in this class except the few recently added tests. I guess it's because the 2nd argument of the as() method tells what we get from it. I agree that reading lines is easier when there is an explicit return type and I fixed my test. Maybe we can refactor a class in the future to bring it to a consistent style.

leontyevdv commented:
> We need some CSV tests for error cases. What happens if you use `TBUCKET` on a data set with no `@timezone` field, for example? Presumably that fails in some graceful way. What other edge cases can we test for?

While writing these CSV tests, I thought about negative tests, but I'm a bit confused about how to write them properly, if they are even needed. For example:

- Missing field edge case. This has already been tested in AnalyzerTests#testLookupMissingField. When a field is missing, we don't return it as a warning, but rather throw an exception. I ran a test with the BUCKET function on my local environment:

```
FROM kibana_sample_data_logs
| STATS logs_per_month = COUNT(*) BY month = BUCKET(start_date, 1 year) 
| SORT month
```

In this query, the start_date doesn't exist, so Kibana displays an exception. The same is true for periods/durations that don't exist or are negative or zero. It seems that these behaviors are system-wide and we don't test them in the function-specific tests.

WDYT?

fang-xing-esql commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/grouping/TBucket.java:
```diff
@@ -0,0 +1,130 @@
+    private final Expression timestamp;
+
+    @FunctionInfo(
+        returnType = { "date" },
```
`Bucket` supports `date_nanos`, do we want `TBucket` to support `date_nanos`?

leontyevdv replied:
Right! I thought that the @timestamp field should be always of date type but I learnt that this can be changed via mapping. I added support of date_nanos, added the corresponding unit tests for that, tests for the unsupported type of the @timestamp field.

fang-xing-esql commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/AnalyzerTests.java:
```diff
@@ -4132,6 +4133,41 @@ public void testGroupingOverridesInInlinestats() {
             """, "Found 1 problem\n" + "line 2:49: Unknown column [x]", "mapping-default.json");
     }
 
+    public void testTBucketWithIntervalInStringInBothAggregationAndGrouping() {
```
Some more tests coverage that I can think of, that are nice to have:
- `LogicalPlanOptimizerTests` that validate that `TBucket` is transformed to `Bucket` successfully.
- Tests that show the error when an index does not have a `@timestamp` field.
- Tests that show the error when the `@timestamp` field's data type is not supported.
- Tests that show the error when the intervals provided are not valid intervals.

fang-xing-esql commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/tbucket.csv-spec:
```diff
@@ -0,0 +1,204 @@
+// TBUCKET-specific tests
```
`Bucket` can also be referenced outside of the `BY` clause in the `STATS` command, can we have some similar tests for `TBucket` to validate the transformation happens correctly? 
```
reuseGroupingFunction#[skip:-8.13.99, reason:BUCKET renamed in 8.14]
FROM employees
| STATS sum = 1 + BUCKET(salary, 1000.) BY b1k = BUCKET(salary, 1000.)
| SORT sum
| LIMIT 4
;
```

leontyevdv replied:
It still didn't work due to the implicit nature of the @timestamp field and inability to resolve such fields. I found a workaround for that by creating a set of functions that require implicit timestamps, then recursively moving through the plan, checking if the functions require a timestamp and adding it to the nameList. See the changes in the class [FieldNameUtils](https://github.com/elastic/elasticsearch/pull/131449/files#diff-6a4e76234dd9ef3b36000cd030ae7546ddabe9283c4715be878a9edba44cc648R62). Please, let me know about what do you think about this change and if there are better ways to implement this.

fang-xing-esql reviewed (commented):
Thank you @leontyevdv! I added some comments about the edge cases that I can think of.

leontyevdv committed:
ES|QL: Address PR review comments

- Add a new test index to AbstractLogicalPlanOptimizerTests to test with
 TBucket
- Add tbucket to bucket translation test
- Add date_nano support to TBucket

Closes #131068

Commit:
[CI] Auto commit changes from spotless

leontyevdv committed:
Merge branch 'main' into feature/esql-tbucket-function

leontyevdv committed:
Merge branch 'main' into feature/esql-tbucket-function

leontyevdv committed:
Merge branch 'main' into feature/esql-tbucket-function

leontyevdv committed:
ES|QL: Address PR review comments

- Add more tests (in TS mode and in Standard mode)
- Add TestClassSignatureTypes to fix docs generation when there are
more than on constructors and not all parameters are exposed publicly.
- Improve TBucket types resolution
- Add SubstituteSurrogateExpressions before TranslateTimeSeriesAggregate
 to substitute TBucket by Bucket before it is used in
 TimeSeriesAggregate

Closes #131068

leontyevdv committed:
ES|QL: Address PR review comments

Fix IT

Closes #131068

leontyevdv committed:
ES|QL: Fix test

Fix an ES|QL test query.

leontyevdv committed:
ES|QL: Fix test

Fix an ES|QL test query.

leontyevdv committed:
Merge branch 'main' into feature/esql-tbucket-function

leontyevdv commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:
```diff
@@ -142,6 +142,7 @@ protected static Batch<LogicalPlan> substitutions() {
             new ReplaceAggregateAggExpressionWithEval(),
             // lastly replace surrogate functions
             new SubstituteSurrogateAggregations(),
+            new SubstituteSurrogateExpressions(),
```
I added this to make TBucket transform to Bucket before it's used in TranslateTimeSeriesAggregate.

fang-xing-esql replied:
I’m wondering why this rule is added here. If you could provide an example of why it’s needed, that would be great, we can then see if it can be avoided. Since this rule is already inside the `substitutions` batch, I’m curious why it needs to be called twice.

fang-xing-esql replied:
If `SubstituteSurrogateExpressions` is required before `TranslateTimeSeriesAggregate`. I wonder if moving `TranslateTimeSeriesAggregate` after `SubstituteSurrogateExpressions` is an option here, instead of adding a duplicated `SubstituteSurrogateExpressions` before it?

alex-spies replied:
Ok, I think the issue is [here](https://github.com/elastic/elasticsearch/blob/6d5099e8c52c2e4942ab45e1b224354fc308fb75/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/TranslateTimeSeriesAggregate.java#L223) in `TranslateTimeSeriesAggregate`. We're only looking for `Bucket` instances and don't take into account substitutions.

Maybe a more localized approach would just check for substitutes here, rather than fully try to run `SubstituteSurrogateExpressions`?

leontyevdv committed:
ES|QL: Polish the code.

Polish the TBUCKET related code.

leontyevdv committed:
Merge branch 'main' into feature/esql-tbucket-function

leontyevdv committed:
ES|QL: Polish the code.

Polish the TBUCKET related code.

leontyevdv commented:
> Does `TBUCKET` also support receiving a start and end time? I think that would be important for typical use cases in Kibana where a specific time range is selected. For example:
> 
> ```
> TBUCKET(?_tstart, ?_tend)
> TBUCKET(50, ?_tstart, ?_tend)
> ```
> 
> See also https://www.elastic.co/docs/explore-analyze/query-filter/languages/esql-kibana#_custom_time_parameters

For that one should use WHERE-BETWEEN clause, if I understood [this](https://www.elastic.co/docs/reference/query-languages/esql/functions-operators/grouping-functions#esql-bucket) correctly. 

<img width="853" height="380" alt="image" src="https://github.com/user-attachments/assets/1e2ddc31-d947-4876-9bc3-c72cf537b1bb" />

leontyevdv committed:
ES|QL: Polish the code.

Add more CsvTests for TBUCKET.

leontyevdv committed:
Merge branch 'main' into feature/esql-tbucket-function

leontyevdv committed:
ES|QL: Add timestamp injection

leontyevdv committed:
Merge branch 'main' into feature/esql-tbucket-function

# Conflicts:
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlCapabilities.java

leontyevdv committed:
ES|QL: Add timestamp injection

leontyevdv commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/tbucket.csv-spec:
```diff
@@ -0,0 +1,204 @@
+
+min:datetime             | max:datetime             | bucket:datetime
+2023-10-23T12:15:03.360Z | 2023-10-23T13:55:01.543Z | 2023-01-01T00:00:00.000Z
+;
```
Yes, it is expected. I've added a test to the [union_types.csv-spec](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql/qa/testFixtures/src/main/resources/union_types.csv-spec#L2502). Thank you!

fang-xing-esql commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
```diff
@@ -8744,4 +8744,79 @@ public List<Attribute> output() {
     }
 
+    public void testTranslateDataGroupedByTBucket() {
+        assumeTrue("requires snapshot builds", Build.current().isSnapshot());
```
Is `tbucket` expected to be behind snapshot? If so, the `TBUCKET` in `ESQLCapability` need to marked with `Build.current().isSnapshot()`. And please also check `isSnapshot()` in the other tests and optimizer code, add `test-release` label can help capture potential release tests issues.

fang-xing-esql reviewed (commented):
Thank you for adding more tests @leontyevdv, and fixing the potential issues they exposed! I have a couple of questions, one is regarding whether `tbucket` is expected to be behind snapshot, the other is about the duplication of `SubstituteSurrogateExpressions` in `LogicalPlanOptimizer` - whether putting the `TranslateTimeSeriesAggregate` and the existing `SubstituteSurrogateExpressions` is an option here.

leontyevdv committed:
Merge branch 'main' into feature/esql-tbucket-function

# Conflicts:
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlCapabilities.java
#	x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/VerifierTests.java

leontyevdv committed:
Merge branch 'main' into feature/esql-tbucket-function

# Conflicts:
#	x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/action/LookupJoinTypesIT.java
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlCapabilities.java
#	x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java
#	x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/DocsV3Support.java
#	x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/DocsV3SupportTests.java
#	x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/predicate/operator/CastOperatorTests.java

leontyevdv committed:
ES|QL: Refactor tests to

- Use TypeSignature instead of Map.Entry
- Refactor LogicalPlanOptimizerTests to use Capabilities

leontyevdv committed:
ES|QL: Refactor tests to

- Use TypeSignature instead of Map.Entry
- Refactor LogicalPlanOptimizerTests to use Capabilities

leontyevdv committed:
ES|QL: Refactor tests to

- Use TypeSignature instead of Map.Entry
- Refactor LogicalPlanOptimizerTests to use Capabilities

leontyevdv committed:
ES|QL: Refactor tests to

- Use TypeSignature instead of Map.Entry
- Refactor LogicalPlanOptimizerTests to use Capabilities

leontyevdv committed:
Merge branch 'main' into feature/esql-tbucket-function

leontyevdv committed:
Merge branch 'main' into feature/esql-tbucket-function

alex-spies reviewed (commented):
Heya, I'm only chiming in with regard to the proposed change of the optimizer rules; didn't consider the rest of the PR.

I'd like to look into how we can avoid another copy of `SubstituteSurrogateExpressions` as additional rules add complexity to the optimizer and are difficult to refactor later.

leontyevdv committed:
ES|QL: Refactor tests to

- Remove SubstituteSurrogateExpressions rule from LogicalPlanOptimizer
- Add TBucket translation to TranslateTimeSeriesAggregate

leontyevdv commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/TranslateTimeSeriesAggregate.java:
```diff
@@ -225,6 +226,12 @@ LogicalPlan translate(TimeSeriesAggregate aggregate) {
                         throw new IllegalArgumentException("expected at most one time bucket");
                     }
                     timeBucketRef.set(e);
+                } else if (child instanceof TBucket tbucket && tbucket.field().equals(timestamp.get())) {
```
@alex-spies , @fang-xing-esql I've removed the duplicating rule from LogicalPlanOptimizer in favor of this piece of code. Please, take a look. Thank you!

alex-spies replied:
Looks more localized now, thank you.

You could attempt the substitution before checking if the substitution result is a `Bucket`, but the current version should work, too.

leontyevdv committed:
Merge branch 'main' into feature/esql-tbucket-function

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/session/FieldNameUtilsTests.java:
```diff
@@ -2985,6 +2985,12 @@ public void testForkBeforeGrok() {
+    public void testImplicitFieldNames() {
+        assertFieldNames("""
+            FROM sample_data
+            | STATS x = 1 year + TBUCKET(1 day) BY b1d = TBUCKET(1 day)""", Set.of("@timestamp", "@timestamp.*"));
```
Hey, ideally let's add a bunch of other queries like this.
Interesting examples use e.g. `KEEP @timestamp` before the `STATS`, or a `KEEP @*` or `KEEP *stamp*`.

Is it valid to have another `STATS` later if @timestamp` survives? Like `STATS ... BY TBUCKET(1 day), @timestamp | WHERE ... | STATS BY TBUCKET(1 hour)`?

Also, what happens if theres an eval `FROM sample_data | EVAL @timestamp = "2024-01-01"::date | STATS ... BY TBUCKET(2 days)`?

For these tests to be robust, we want to be as creative as possible :)

leontyevdv replied:
Hey Alex, I added all the tests that you mentioned to both FieldNameUtilsTests and CsvTests. Thanks!

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/FieldNameUtils.java:
```diff
@@ -166,6 +171,13 @@ public static PreAnalysisResult resolveFieldNames(LogicalPlan parsed, EnrichReso
                 }
+
+                p.forEachExpression(UnresolvedFunction.class, uf -> {
+                    if (FUNCTIONS_REQUIRING_TIMESTAMP.contains(uf.name().toLowerCase(Locale.ROOT))) {
```
This change looks correct to me, but I'd like to solicit a review by @astefan just for this specific part of the PR as this is delicate when done wrong.

astefan replied:
I am looking now at this. Thank you for the ping @alex-spies

leontyevdv committed:
Merge branch 'main' into feature/esql-tbucket-function

leontyevdv committed:
ES|QL: Improve tests

- Add more tests for corner cases

leontyevdv committed:
ES|QL: Improve tests

- Fix IT by adding SORT

leontyevdv committed:
ES|QL: Improve TBucket

leontyevdv committed:
Merge branch 'main' into feature/esql-tbucket-function

# Conflicts:
#	x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/VerifierTests.java

leontyevdv committed:
Merge branch 'main' into feature/esql-tbucket-function

leontyevdv committed:
ES|QL: Improve tests

fang-xing-esql commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/grouping/TBucket.java:
```diff
@@ -0,0 +1,141 @@
+        this.timestamp = timestamp;
+    }
+
+    private TBucket(StreamInput in) throws IOException {
+        this(Source.readFrom((PlanStreamInput) in), in.readNamedWriteable(Expression.class), in.readNamedWriteable(Expression.class));
+    }
```
This can be removed if no sterilization is needed.

not-napoleon reviewed (approved):
Great work, sorry it's been more complicated than expected.

fang-xing-esql reviewed (approved):
Thank you @leontyevdv ! I added one last comment related to serialization of `tbucket`, the rest LGTM.

leontyevdv committed:
ES|QL: Improve tests

astefan reviewed (commented):
I am wondering why do we consider such a scenario an acceptable one:

```
from test | stats max(emp_no) by tbucket(1 hour)`
```

generating an error like `Unknown column [@timestamp]`.
Do we have another case in ESQL where the user is supposed to know that `@timestamp` is a field that must be present somewhere even though the user **didn't actually type in the query** `@timestamp`?

Imho, this would be an **acceptable** use case if the query would be `TS test | stats ...... by tbucket(1 hour)` meaning the user is aware that by using `TS` source command, it is expected to be in the area of "timeseries" indices and queries and some things (like the `@timestamp` field presence) are somewhat expected to happen.

---

Also, if I run this query `from employees | stats min(salary) by tbucket(birth_date)` I get back an error message `Unknown column [@timestamp]`.

**It is ok** when running `..... by tbucket(birth_date,1month)` to get back `ql_illegal_argument_exception expects exactly one argument` but, as an user who is exploring `tbucket`, if I remove `1month` and keep `birth_date` (which is a `date` field) to get back something that has nothing to do with my query, it is unexpected.

@martijnvg @kkrik-es @dnhatn thoughts?

kkrik-es commented:
@astefan we expect `tbucket` to apply to all data streams that implicitly define `@timestamp`. This is not limited to metrics,  should be applicable to logs and more. I'm not sure if we have such a precedent in ESQL, but we should try to simplify the syntax for such heavily used applications imho.

astefan reviewed (approved):
LGTM. My earlier comment is unrelated to what, technically, the PR is doing. Please, regard that comment as an observation and something to discuss post-merge, if my observation is valid.

leontyevdv commented:
Thank you all for your feedback, folks! I will definitely address all the suggestions for improvements in the following PR since this one is getting harder to maintain and to follow because of its size and the number of discussions.
